### DISCUSSION TITLE
Enforce value of an object property as key on a record?

### DISCUSSION THREAD

**Discussion by codepunkt** [colinhacks/zod#1337] [Q&A] [answered]
I'm looking to validate something like this - is this possible with zod?

```ts
const options = z.object({
  defaultContentLanguage: z.string(),
  languages: z.record(z.string(), z.object({})) // must include the value of `defaultContentLanguage` as a record key
})
```

**Comment by scotttrinh** [comment, marked as answer]
You can validate it at runtime, but you'll end up with `Record<string, string>` since TypeScript cannot know what the value will be at runtime. If `defaultContentLanguage` is some kind of enum, you could at least get _that_ checked at compile-time, but it'll just be a partial record since, again, we can't know at compile-time which value is passed at run-time.

```ts
const options = z.object({
  defaultContentLanguage: z.string(),
  languages: z.record(z.string())
}).refine(opt => Object.keys(opt.languages).includes(opt.defaultContentLanguage))
```

**Reply by codepunkt** [reply]
Checks like `options.parse(opts)` will be performed at runtime. It would be sufficient for me to check then.
Can a validation like that be done with `zod` or is the sole premise to check things at runtime that can also be made type-safe at compile-time?

**Reply by scotttrinh** [reply]
The code I provided will do the runtime check, so it should be appropriate for your use case. As long as you're fine with `languages` being `Record<string, string>`, this will accomplish the runtime behavior you want.
